analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs

```diff
@@ -128,14 +129,122 @@ public static SyntaxNode WalkUpParentheses(this SyntaxNode node)
+        public static SyntaxNode FindAssignmentComplement(this SyntaxNode node)
         {
+            var thisSide = node.Ancestors()
+                .TakeWhile(x => x.IsAnyKind(
+                    SyntaxKind.Argument,
+                    SyntaxKindEx.TupleExpression,
+                    SyntaxKindEx.SingleVariableDesignation,
+                    SyntaxKindEx.ParenthesizedVariableDesignation,
+                    SyntaxKindEx.DiscardDesignation,
+                    SyntaxKindEx.DeclarationExpression))
+                .LastOrDefault();
+            if ((TupleExpressionSyntaxWrapper.IsInstance(thisSide) || DeclarationExpressionSyntaxWrapper.IsInstance(thisSide))
+                && thisSide.Parent is AssignmentExpressionSyntax assignment)
+            {
+                var otherSide = assignment switch
+                {
+                    { Left: { } left, Right: { } right } when left.Equals(thisSide) => right,
+                    { Left: { } left, Right: { } right } when right.Equals(thisSide) => left,
+                    _ => null,
+                };
+                if (TupleExpressionSyntaxWrapper.IsInstance(otherSide) || DeclarationExpressionSyntaxWrapper.IsInstance(otherSide))
+                {
+                    var indexAndCount = IndexAndCountOfNesting(node);
+                    return FindMatchingNestedNode(indexAndCount, otherSide);
+                }
+            }
+
+            return null;
+
+            static Stack<IndexCountPair> IndexAndCountOfNesting(SyntaxNode node)
+            {
+                Stack<IndexCountPair> indexAndCount = new();
+                while (TupleExpressionSyntaxWrapper.IsInstance(node?.Parent) || ParenthesizedVariableDesignationSyntaxWrapper.IsInstance(node?.Parent))
+                {
+                    node = node switch
+                    {
+                        ArgumentSyntax tupleArgument when TupleExpressionSyntaxWrapper.IsInstance(node.Parent) =>
+                            PushIndexAndCountTuple(indexAndCount, (TupleExpressionSyntaxWrapper)node.Parent, tupleArgument),
+                        _ when VariableDesignationSyntaxWrapper.IsInstance(node) && ParenthesizedVariableDesignationSyntaxWrapper.IsInstance(node.Parent) =>
+                            PushIndexAndCountParenthesizedDesignation(indexAndCount, (ParenthesizedVariableDesignationSyntaxWrapper)node.Parent, (VariableDesignationSyntaxWrapper)node),
+                        _ => null,
+                    };
+                    if (DeclarationExpressionSyntaxWrapper.IsInstance(node?.Parent) && node is { Parent.Parent: ArgumentSyntax { } argument })
+                    {
+                        node = argument;
+                    }
+                }
+                return indexAndCount;
+            }
+
+            static SyntaxNode FindMatchingNestedNode(Stack<IndexCountPair> indexAndCount, SyntaxNode node)
+            {
+                while (node is not null && indexAndCount.Count > 0)
```

Comment:
better not to change argument _and_ improve naming
```suggestion
                var matchedNestedNode = otherOutermostNode;
                while (matchedNestedNode is not null && indexAndCount.Count > 0)
```